readme.md
# Qonquest
Turn-based strategy game in the terminal.
## Build instructions
Get terminal.d [here](https://github.com/adamdruppe/arsd/blob/master/terminal.d), create a new folder "arsd", and put it in that folder. Then simply run `dub build`.

changelog.txt
Additions:
* Added toggles
Fixes:
* Fixed a bug where AI would run even if they have no provinces
* Added some miscallenous optimizations

source/map.d
module map; /// handles map loading, drawing, and holds map-related classes

import script, app, action;
import std.file, std.typecons, std.stdio, std.string, std.traits, std.typecons, std.random;
import arsd.terminal;

ushort mapWidth;
ushort mapHeight;

/// Represents a single province in the map.
struct Province {
  Tuple!(ushort, ushort)[] tiles; /// Every tile in the province.
  bool ocean = false; /// Whether the province is ocean or not.
  string name = ""; /// The name of the province.
  Country* owner; /// Pointer to the country that owns the province.
  ubyte color = 0; /// The color of the province. (see: qonquest.app.color)
  Tuple!(ushort, ushort) city; /// Location of the city of the province.
  int[string] vars; /// Variables for the province. (currently unused)
  int troops = 0; /// The number of troops in the province.
  ushort[] adjacencies; /// The provinces that are adjacent to this one.
  char ch = ' '; /// The character to draw for the province.
}
Province[ushort] provinces; /// List of all provinces in the map.
/// Finds a province by name
Tuple!(Province*, ushort) findProvince(string name) {
  foreach(ushort i, ref Province p; provinces) {
    if(p.name == name) return tuple(&p, i);
  }
  return Tuple!(Province*, ushort)(null, 0);
}

/// Represents a single country in the map.
class Country {
  string name; /// The name of the country.
  ubyte color; /// The color of the country. (see: qonquest.app.color)
  ushort capital; /// The province that is the capital of the country.
  int[string] vars; /// Variables for the country. (currently unused)
  char ch = ' '; /// The character to draw for the country.
  this() {}

  Province[] getAllProvinces() {
    /// Returns all provinces that are owned by the country.
    Province[] res;
    foreach(Province p; provinces)
      if(p.owner != null && *(p.owner) == this)
        res ~= p;
    return res;
  }

  /// Returns the IDs of all provinces that are owned by the country.
  ushort[] getAllProvinceIDs() {
    ushort[] res;
    foreach(ushort i, Province p; provinces)
      if(p.owner != null && *(p.owner) == this)
        res ~= i;
    return res;
  }

  /// Returns the amount of troops gained per turn
  int troopsPerTurn() {
    return cast(int)(getAllProvinces().length*2);
  }

  /**
    Returns all provinces on the border of this country as a tuple of
    [0] = province from (owned by this country)
    [1] = province to (owned by other country)
  */
  Tuple!(Province*, Province*)[] frontiers() {
    Tuple!(Province*, Province*)[] res;
    foreach(ref Province p; provinces) {
      foreach(ushort j; p.adjacencies) {
        if(p.owner != null && *(p.owner) == this && provinces[j].owner != null && *(provinces[j].owner) != this) {
          res ~= tuple(&p, &(provinces[j]));
        }
      }
    }
    return res;
  }

  /// Runs the AI for the country.
  void ai(Terminal* t) {
    t.write("Running AI for ");
    t.color(colors[color], Color.black);
    t.writefln("%s", name);
    t.color(Color.white, Color.black);

    int troopsRemaining = troopsPerTurn();
    auto frontiers = frontiers();
    auto rnd = Random(unpredictableSeed);
    // distribute troops
    while(troopsRemaining > 0 && frontiers.length > 0) {
      int idx = rnd.front%cast(int)(frontiers.length);
      rnd.popFront();
      frontiers[idx][0].troops++;
      troopsRemaining--;
    }
    if(turn > 1) {
      Action[] actions;
      foreach(f; frontiers) {
        if(f[0].troops != 0) { // if we have troops to move
          if(f[0].troops >= f[1].troops) { // if we have more troops than the other country
            actions ~= new MoveAction(f[0], f[1]); // attack
          }
        }
      }
      commitAll(actions, t);
    }
  }

  /// Updates the country.
  void update(Terminal* t) {
    ushort[] provs = getAllProvinceIDs();
    auto rnd = Random(unpredictableSeed);
    if(*(provinces[capital].owner) != this && provs.length > 0) {
      // capital has been conquered
      // set new capital to random owned province
      ushort newCap = provs[rnd.front%cast(int)(provs.length)];
      capital = newCap;
    }
  }
}
Country[string] countries; /// List of all countries in the map.
/// Finds a country by name
Country* findCountry(string name) {
  foreach(ref Country c; countries) {
    if(c.name == name) return &c;
  }
  return null;
}
/// Returns a tag of a country, given a pointer to the country. Returns "" if the country can not be found.
string tagOf(Country* c) {
  foreach(string tag, ref Country c2; countries) {
    if(c == &c2) return tag;
  }
  return "";
}
/// Represents a strait on the map
struct Strait {
  Tuple!(ushort, ushort) from;
  Tuple!(ushort, ushort) to;
}
Strait[] straits; /// List of all straits in the map.

/// Loads provinces from a map file (generated by mapgen.py)
void loadMap(string file, Terminal* t) {
  Province[ushort] empty;
  provinces = empty; // reset provinces. idk if this is how you're supposed to do it, but it works
  ubyte[] data = cast(ubyte[])std.file.read(file);
  int dataStart = data[0] | (data[1] << 8); // where the map data starts
  mapWidth = data[2] | (data[3] << 8); // width of the map
  mapHeight = data[4] | (data[5] << 8); // height of the map
  for(ushort i = 0; i < mapWidth*mapHeight; i++) {
    ushort x = i%mapWidth;
    ushort y = i/mapWidth;
    ushort id = data[dataStart+i*2] | (data[dataStart+i*2+1] << 8);
    if(id in provinces) {
      // add tile to province
      provinces[id].tiles ~= tuple(x, y);
    } else {
      // create new province and run province setup
      provinces[id] = Province([tuple(x, y)]);
      runScript(format("./data/scripts/provinces/%d.qsc", id), new ProvinceScope(&(provinces[id])), t);
    }
  }
}

// for some reason I need 2 different methods for creating a 2d array
// the first acts on primitives (ints, floats, etc)
// the second acts on everything else
T[][] create2dArrayPrim(T)(int width, int height) {
  T[][] arr;
  for(int i = 0; i < width; i++) {
    T[] row;
    for(int j = 0; j < height; j++) {
      row ~= *(new T());
    }
    arr ~= row;
  }
  return arr;
}
T[][] create2dArray(T)(int width, int height) {
  T[][] arr;
  for(int i = 0; i < width; i++) {
    T[] row;
    for(int j = 0; j < height; j++) {
      row ~= new T();
    }
    arr ~= row;
  }
  return arr;
}

/// returns a 2D array of ushorts, where each element is the ID of the province at that location
ushort[][] constructMap() {
  ushort[][] map = create2dArrayPrim!ushort(mapWidth, mapHeight);
  foreach(ushort id, Province p; provinces) {
    foreach(Tuple!(int, int) t; p.tiles) {
      map[t[0]][t[1]] = id;
    }
  }
  return map;
}

enum MapRenderType {
  PROVINCE, COUNTRY
}

class Tile {
  Tuple!(int, int) col;
  dchar c;
  this() {
    c = ' ';
    col = tuple(Color.white, Color.black);
  }
  this(dchar c, Tuple!(int, int) col) {
    this.c = c;
    this.col = col;
  }
}

void renderTiles(Tile[][] tiles, int w, int h, Terminal* t) {
  for(int i = 0; i < h; i++) {
    for(int j = 0; j < w; j++) {
      Tile tile = tiles[j][i];
      t.color(tile.col[0], tile.col[1]);
      t.write(tile.c);
    }
    t.writeln();
  }
  t.color(Color.white, Color.black);
}

/// Renders the map
void renderMap(MapRenderType mrt, Terminal* t) {
  ushort[][] provs = constructMap();
  Tile[][] tiles = create2dArray!Tile(mapWidth, mapHeight);
  void tileWriteColor(int x, int y, string s, Tuple!(int, int) col) {
    for(int i = 0; i < s.length; i++) {
      tiles[x+i][y].c = s[i];
      tiles[x+i][y].col = col;
    }
  }
  void tileWrite(int x, int y, string s, int fg) {
    for(int i = 0; i < s.length; i++) {
      if(x+i >= mapWidth) continue;
      tiles[x+i][y].c = s[i];
      tiles[x+i][y].col[0] = fg;
    }
  }
  final switch(mrt) {
    case MapRenderType.PROVINCE:
      for(int y = 0; y < mapHeight; y++) {
        for(int x = 0; x < mapWidth; x++) {
          Province p = provinces[provs[x][y]];
          if(p.ocean) {
            tiles[x][y] = new Tile('~', tuple(colors[p.color], cast(int)(Color.black)));
          } else {
            tiles[x][y] = new Tile(p.ch, tuple(cast(int)(Color.white), colors[p.color]));
          }
          
        }
      }
      foreach(Province p; provinces) {
        if(p.ocean) continue;
        tileWriteColor(p.city[0], p.city[1], "@", tuple(cast(int)(Color.white), colors[p.color]));
        string s = format("%s, %s", p.name, p.owner.name);
        int x = cast(int)(p.city[0]-(s.length/2));
        tileWrite(x, p.city[1]+1, s, cast(int)Color.white);
        s = format("%d", p.troops);
        x = cast(int)(p.city[0]-(s.length/2));
        tileWrite(x, p.city[1], s, cast(int)Color.white);
      }
      break;
    case MapRenderType.COUNTRY:
      for(int y = 0; y < mapHeight; y++) {
        for(int x = 0; x < mapWidth; x++) {
          Province p = provinces[provs[x][y]];
          if(p.ocean) {
            tiles[x][y] = new Tile('~', tuple(colors[p.color], cast(int)(Color.black)));
          } else {
            if(p.owner != null) tiles[x][y] = new Tile(p.owner.ch, tuple(cast(int)(Color.white), colors[p.owner.color]));     
          }
        }
      }
      foreach(Country c; countries) {
        if(c.getAllProvinces().length > 0) { // make sure country exists
          Province capital = provinces[c.capital];
          int x = cast(int)(capital.city[0]-(c.name.length/2));
          tileWrite(x, capital.city[1], c.name, cast(int)(Color.white));
        }
      }
      break;
  }
  // add straits
  foreach(Strait s; straits) {
    ushort x0 = s.from[0];
    ushort y0 = s.from[1];
    ushort x1 = s.to[0];
    ushort y1 = s.to[1];
    void d() {
      if(tiles[x0][y0].c == '~')
        tiles[x0][y0] = new Tile('#', tuple(cast(int)(Color.red | Bright), cast(int)(Color.black)));
    }
    d();
    while(x0 != x1 || y0 != y1) {
      if(x0 < x1) x0++;
      if(x0 > x1) x0--;
      if(y0 < y1) y0++;
      if(y0 > y1) y0--;
      d();
    }
  }
  renderTiles(tiles, mapWidth, mapHeight, t);
  t.color(Color.white, Color.black);
}

/// Registers a country with the game (triggered by the qsc command "registerCountry")
void registerCountry(string tag, Terminal* t) {
  countries[tag] = new Country();
  runScript(format("./data/scripts/countries/%s.qsc", tag), new CountryScope(&(countries[tag])), t);
}

source/save.d
module qonquest.save;

import std.conv, std.zlib, std.file, std.string;
import qonquest.app, qonquest.map, qonquest.script;

ubyte[] toArray(string s) {
  /// Converts a string to an array of bytes.
  ubyte[] res;
  foreach(char c; s) res ~= cast(ubyte)c;
  return res;
}

string fromArray(ubyte[] a) {
  /// Converts an array of bytes to a string.
  string res;
  foreach(ubyte b; a) res ~= cast(char)b;
  return res;
}

ubyte[] save(ushort ver) {
  /// Saves the current game given a version number.
  ubyte[] file;
  file ~= [ver & 0xFF, ver >> 8]; // format (little endian)
  final switch(ver) {
    case 0: {
      file ~= [0, 0]; // data start, (reserving for later)
      file ~= tagOf(player).toArray()~0; // player tag
      ushort data = to!ushort(file.length);
      // store data position
      file[2] = data&0xFF;
      file[3] = (data>>8)&0xFF;
      string script = "";
      script ~= format("turn %d;", turn);
      foreach(ushort id, Province p; provinces) {
        if(p.ocean) continue;
        script ~= format("province %d {", id);
        script ~= format(`owner "%s";`, tagOf(p.owner));
        script ~= format(`troops %d;`, p.troops);
        script ~= "}";
      }
      foreach(string tag, Country c; countries) {
        script ~= format(`country "%s" {`, tag);
        script ~= format(`capital %d;`, c.capital);
        script ~= "}";
      }
      file ~= script.toArray().compress();
    }
  }
  return file;
}

void load(ubyte[] save) {
  /// Loads a game from a save file.
  int ptr = 0;
  ubyte next() {
    return save[ptr++];
  }
  ushort nexts() {
    return next | (next<<8);
  }
  ubyte curr() {
    return save[ptr];
  }
  ushort ver = nexts;
  final switch(ver) {
    case 0: {
      ushort start = nexts;
      string tag = "";
      while(curr != 0) tag ~= cast(char)next;
      string script = (cast(ubyte[])save[start..$].uncompress()).fromArray();
      interpret(parseScript(script), new GlobalScope(), null);
      gameMode = GameMode.GAME;
      player = &(countries[tag]);
      firstTurn = true;
    }
  }
}

string saveList() {
  /// Lists all saved games.
  string s = "";
  foreach(string file; dirEntries("data/saves/", SpanMode.shallow)) {
    if(file.endsWith(".qsf")) {
      string saveName = file[file.lastIndexOf("/")+1..$-4];
      s ~= format("%s\n", saveName);
    }
  }
  return s;
}

source/action.d
module action;

import map, app;
import arsd.terminal;
import std.string, std.random, std.algorithm;
import core.thread;

/// Represents a single action that can be performed by a player or AI.
interface Action {
  void commit(Terminal* t); /// Peforms the action.
  void display(Terminal* t); /// Displays the action.
}

/// Represents a movement.
class MoveAction : Action {
  Province* source; /// source tile
  Province* dest; /// destination tile
  this(Province* source, Province* dest) {
    this.source = source;
    this.dest = dest;
  }
  void commit(Terminal* t) {
    if(source.troops == 0) return;
    if(source.owner == dest.owner) {
      // just move troops if the owner is the same
      dest.troops += source.troops;
      source.troops = 0;
    } else {
      // do battle
      bool display = source.owner == player || dest.owner == player || toggles["seeAllBattles"];
      if(display) {
        t.writefln("Battle of %s", dest.name);
        t.color(colors[source.owner.color], Color.black);
        t.write(source.owner.name);
        t.color(Color.white, Color.black);
        t.write(" vs ");
        t.color(colors[dest.owner.color], Color.black);
        t.writeln(dest.owner.name);
        t.color(Color.white, Color.black);
      }
      for(int round = 1; source.troops > 0 && dest.troops > 0; round++) {
        auto rnd = Random(unpredictableSeed);
        ubyte die() {
          ubyte res = (rnd.front()%6)+1;
          rnd.popFront();
          return res;
        }
        ubyte[] attackerRolls = [die, die].sort!("a > b").release();
        ubyte[] defenderRolls = [die, die, die].sort!("a > b").release();
        int attackerLost = 0;
        int defenderLost = 0;
        if(attackerRolls[0] > defenderRolls[0]) defenderLost++;
        else attackerLost++;
        if(attackerRolls[1] > defenderRolls[1]) defenderLost++;
        else attackerLost++;
        if(display) {
          t.writefln("Round %d:", round);
          t.color(colors[source.owner.color], Color.black);
          t.writef("%d ", source.troops);
          t.color(Color.white, Color.black);
          t.writef("/ ");
          t.color(colors[dest.owner.color], Color.black);
          t.writefln("%d", dest.troops);
          t.color(Color.white, Color.black);
          t.writefln("Attacker rolls: %d, %d", attackerRolls[0], attackerRolls[1]);
          t.writefln("Defender rolls: %d, %d, %d", defenderRolls[0], defenderRolls[1], defenderRolls[2]);
          t.writefln("Attacker lost %d troops, defender lost %d troops", attackerLost, defenderLost);
          t.writeln();
          if(toggles["waitBattleRounds"]) Thread.sleep(dur!("msecs")(500));
        }
        source.troops -= attackerLost;
        dest.troops -= defenderLost;
      }
      if(source.troops > 0) {
        // attacker won, give the province to the attacker
        dest.owner = source.owner;
        dest.troops = source.troops;
        source.troops = 0;
        if(display) {
          t.write("Result: ");
          t.color(colors[source.owner.color], Color.black);
          t.write(source.owner.name);
          t.color(Color.white, Color.black);
          t.writefln(" conquered %s", dest.name);
        }
      } else if(display) {
        t.write("Result: ");
        t.color(colors[source.owner.color], Color.black);
        t.write(source.owner.name);
        t.color(Color.white, Color.black);
        t.writefln(" lost the battle");
      }
      // make sure none are negative
      if(source.troops < 0) source.troops = 0;
      if(dest.troops < 0) dest.troops = 0;
    }
  }
  void display(Terminal* t) {
    t.write("Move troops from ");
    t.color(colors[source.owner.color], Color.black);
    t.writef("%s (%s)", source.name, source.owner.name);
    t.color(Color.white, Color.black);
    t.writef(" to ");
    t.color(colors[dest.owner.color], Color.black);
    t.writef("%s (%s)", dest.name, dest.owner.name);
    t.writeln();
    t.color(Color.white, Color.black);
  }
}

void commitAll(Action[] actions, Terminal* t) {
  /// Commits a list of actions.
  foreach(Action a; actions) {
    a.commit(t);
  }
}

mapgen/mapgen.py
# simple script to generate a map file from an image
# requires PIL
from PIL import Image;
import sys;
img = Image.open("map.png");
w,h = img.size;
map = [];
map += [6, 0] # data start
map += [w&0xFF, (w>>8)]; # width (little endian)
map += [h&0xFF, (h>>8)]; # height (little endian)
dict = {
  # color: province id
  (255, 255, 255, 255): 0,
  (0, 0, 255, 255): 1,
  (255, 255, 0, 255): 2,
  (255, 0, 0, 255): 3,
  (64, 64, 64, 255): 4,
  (0, 255, 0, 255): 5,
  (255, 127, 0, 255): 6,
  (0, 127, 0, 255): 7,
  (127, 0, 0, 255): 8,
  (0, 0, 127, 255): 9,
  (0, 255, 255, 255): 10,
  (87, 0, 127, 255): 11,
  (178, 0, 255, 255): 12,
  (0, 148, 255, 255): 13,
  (255, 0, 110, 255): 14,
  (127, 51, 0, 255): 15,
  (252, 205, 229, 255): 16,
  (128, 128, 128, 255): 17,
  (204, 204, 204, 255): 18,
  (0, 127, 127, 255): 19
}
for y in range(h):
  for x in range(w):
    color = img.getpixel((x,y));
    if color in dict:
      id = dict[color];
      map += [id&0xFF, (id>>8)];
    else:
      map += [0, 0]; # assume province ID 0
map = bytes(map);
with open("map.bin", "wb") as f:
  f.write(map);

source/app.d
module app;

import map, script, action, saveload;
import std.stdio, std.string, std.file, std.conv, std.algorithm, std.typecons;
import core.stdc.stdlib : exit;
import arsd.terminal;

int[] colors = [Color.black, Color.red, Color.green, Color.yellow, Color.blue, Color.magenta, Color.cyan, Color.white, Color.black | Bright, Color.red | Bright, Color.green | Bright, Color.yellow | Bright, Color.blue | Bright, Color.magenta | Bright, Color.cyan | Bright, Color.white | Bright];
/*
For convinience:
0 - black
1 - red
2 - green
3 - yellow
4 - blue
5 - magenta
6 - cyan
7 - white
8 - black | Bright
9 - red | Bright
10 - green | Bright
11 - yellow | Bright
12 - blue | Bright
13 - magenta | Bright
14 - cyan | Bright
15 - white | Bright
*/

enum GameMode {
  /**
    GameMode is used to represent what state the game currently is in.
  **/
  MAIN_MENU, /// Main Menu
  GAME /// Mid-game
}

GameMode gameMode;

/**
  Error contains the string representation of errors that can occur in the game. 
  These are formatted using D's format() function.
**/
enum Error {
  COMMAND_REQUIRES_ARGS = "Command \"%s\" requires exactly %d arguments.",
  COMMAND_REQUIRES_AT_LEAST_ARGS = "Command \"%s\" requires at least %d arguments.",
  NO_COMMAND = "No command \"%s\" found.",
  INVALID_SUBCOMMAND = "Invalid subcommand \"%s\".",
  COUNTRY_DOESNT_EXIST = "Country \"%s\" doesn't exist.",
  PROVINCE_DOESNT_EXIST = "Province \"%s\" doesn't exist.",
  NOT_ENOUGH_TROOPS = "You have %d troops, but are attempting to deploy %d troops.",
  NOT_NUMBER = "\"%s\" is not a number.",
  NOT_DEPLOYED_ALL_TROOPS = "You must deploy all your troops before ending your turn.",
  CANT_MOVE_TURN_1 = "You cannot move troops on turn 1.",
  PROVINCE_NOT_OWNED = "You do not own %s.",
  NOT_ADJACENT = "%s is not adjacent to %s.",
  NO_TOGGLE = "Toggle \"%s\" does not exist.",
  INVALID_AMOUNT = "Invalid amount %s.",
}

/// CommandException is used to represent errors that occur when a command is executed.
class CommandException : Exception {
  this(string message) {
    super(message);
  }
}

Country* player; /// Pointer to the country the player is playing as
Action[] actions; /// List of actions that have been performed by the player this turn
int turn = 0; /// The turn number
int troopsToDeploy = 0; /// The number of troops the player has to deploy this turn
bool[string] toggles; /// List of toggles that may be enabled or disabled
bool firstTurn = true; /// Whether or not the player is on the first turn (this may seem redundant, but it's important when saves are loaded)

void main() {
  auto t = Terminal(ConsoleOutputType.linear);
  t.color(Color.white, Color.black);
  t.clear();
  try {
    loadScripts("./data/scripts", &t);
    runScript("./data/scripts/launch.qsc", new GlobalScope(), &t);
    loadMap("./data/map.bin", &t);
    changeMode(GameMode.MAIN_MENU, &t);
    toggles["waitBattleRounds"] = false;
    toggles["seeAllBattles"] = false;
    toggles["cheats"] = false;
    void expectArgs(string cmd, string[] args, int n) {
      if(args.length != n) {
        throw new CommandException(format(Error.COMMAND_REQUIRES_ARGS, cmd, n));
      }
    }
    void expectAtLeastArgs(string cmd, string[] args, int n) {
      if(args.length < n) {
        throw new CommandException(format(Error.COMMAND_REQUIRES_AT_LEAST_ARGS, cmd, n));
      }
    }
    while(true) {
      try {
        final switch(gameMode) {
          case GameMode.MAIN_MENU: {
            t.write("$ ");
            string[] command = nextCommand(&t);
            if(command.length > 0) {
              string cmd = command[0];
              string[] args = command[1..$];
              switch(cmd) {
                case "help":
                  expectArgs(cmd, args, 0);
                  t.writeln("Main menu commands:");
                  t.writeln("  help\n\tdisplays this help message");
                  t.writeln("  map <province/country>\n\tdisplays the map");
                  t.writeln("  play <country>\n\tstarts a game as <country>");
                  t.writeln("  changelog\n\tdisplays the changelog");
                  t.writeln("  load <name>\n\tloads a saved game");
                  t.writeln("  saves\n\tdisplays a list of saved games");
                  t.writeln("  quit\n\tquits the game");
                  break;
                case "map":
                  expectArgs(cmd, args, 1);
                  switch(args[0]) {
                    case "province":
                      renderMap(MapRenderType.PROVINCE, &t);
                      break;
                    case "country":
                      renderMap(MapRenderType.COUNTRY, &t);
                      break;
                    default:
                      throw new CommandException(format(Error.INVALID_SUBCOMMAND, args[0]));
                  }
                  break;
                case "play": {
                  expectArgs(cmd, args, 1);
                  Country* c = findCountry(args[0]);
                  if(c is null) throw new CommandException(format(Error.COUNTRY_DOESNT_EXIST, args[0]));
                  player = c;
                  changeMode(GameMode.GAME, &t);
                  break;
                }
                case "changelog": {
                  expectArgs(cmd, args, 0);
                  t.writeln("Changelog:");
                  string[] changelog = readText("./data/changelog.txt").splitLines();
                  foreach(string s; changelog) {
                    if(s.indexOf("+") != -1) {
                      t.color(Color.green | Bright, Color.black);
                    } else if(s.indexOf("-") != -1) {
                      t.color(Color.red | Bright, Color.black);
                    } else if(s.indexOf("~") != -1) {
                      t.color(Color.yellow | Bright, Color.black);
                    } else {
                      t.color(Color.white, Color.black);
                    }
                    t.writeln(s);
                  }
                  t.color(Color.white, Color.black);
                  break;
                }
                case "load":
                  expectArgs(cmd, args, 1);
                  try {
                    load(cast(ubyte[])read(format("./data/saves/%s.qsf", args[0])));
                    nextTurn(&t);
                  } catch(FileException e) {
                    writefln("\nError loading save: %s", e.message);
                  }
                  break;
                case "saves":
                  expectArgs(cmd, args, 0);
                  t.writeln(saveList());
                  break;
                case "quit":
                  expectArgs(cmd, args, 0);
                  return;
                default:
                  throw new CommandException(format(Error.NO_COMMAND, cmd));
              }
            }
            break;
          }
          case GameMode.GAME: {
            t.writef("%s$ ", player.name);
            string[] command = nextCommand(&t);
            if(command.length > 0) {
              string cmd = command[0];
              string[] args = command[1..$];
              switch(cmd) {
                case "help":
                  expectArgs(cmd, args, 0);
                  t.writeln("Game commands:");
                  t.writeln("  help\n\tdisplays this help message");
                  t.writeln("  map <province/country>\n\tdisplays the map");
                  t.writeln("  quit\n\tgoes back to the main menu");
                  t.writeln("  actions\n\tdisplays all actions taken in this turn");
                  t.writeln("  move <source> <dest>\n\tMove all units from <source> to <dest>. The provinces must be adjacent.");
                  t.writeln("  deploy <amt> <province>\n\tDeploy <amt> troops to <province>");
                  t.writeln("  end\n\tends the turn");
                  t.writeln("  toggle\n\ttoggles a toggle");
                  t.writeln("  toggles\n\tdisplays all toggles");
                  t.writeln("  save <name>\n\tsaves the game to file <name>");
                  if(toggles["cheats"]) {
                    t.writeln("Cheat commands:");
                    t.writeln("  script <script>\n\texecutes a script");
                    t.writeln("  run <command>\n\truns the input as a script");
                  }
                  break;
                case "map":
                  expectArgs(cmd, args, 1);
                  switch(args[0]) {
                    case "province":
                      renderMap(MapRenderType.PROVINCE, &t);
                      break;
                    case "country":
                      renderMap(MapRenderType.COUNTRY, &t);
                      break;
                    default:
                      throw new CommandException(format(Error.INVALID_SUBCOMMAND, args[0]));
                  }
                  break;
                case "quit":
                  expectArgs(cmd, args, 0);
                  changeMode(GameMode.MAIN_MENU, &t);
                  break;
                case "actions":
                  expectArgs(cmd, args, 0);
                  foreach(Action a; actions) a.display(&t);
                  break;
                case "move": {
                  if(turn == 1) throw new CommandException(Error.CANT_MOVE_TURN_1);
                  expectArgs(cmd, args, 2);
                  Tuple!(Province*, ushort) p0 = findProvince(args[0]);
                  Tuple!(Province*, ushort) p1 = findProvince(args[1]);
                  Province* source = p0[0];
                  Province* dest = p1[0];
                  ushort destId = p1[1];
                  if(source is null) throw new CommandException(format(Error.PROVINCE_DOESNT_EXIST, args[0]));
                  if(source.owner != player) throw new CommandException(format(Error.PROVINCE_NOT_OWNED, args[0]));
                  if(dest is null) throw new CommandException(format(Error.PROVINCE_DOESNT_EXIST, args[1]));
                  if(!source.adjacencies.canFind(destId))
                    throw new CommandException(format(Error.NOT_ADJACENT, args[0], args[1]));
                  Action a = new MoveAction(source, dest);
                  t.write("Action added: ");
                  a.display(&t);
                  actions ~= a;
                  break;
                }
                case "deploy": {
                  expectArgs(cmd, args, 2);
                  try {
                    int amt = to!int(args[0]);
                    if(amt == -25) writeln("shoutouts to ibrokqr");
                    if(amt < 0) throw new CommandException(format(Error.INVALID_AMOUNT, args[0]));
                    Province* p = findProvince(args[1])[0];
                    if(p is null) throw new CommandException(format(Error.PROVINCE_DOESNT_EXIST, args[1]));
                    if(amt > troopsToDeploy) 
                      throw new CommandException(format(Error.NOT_ENOUGH_TROOPS, troopsToDeploy, amt));
                    if(p.owner != player) throw new CommandException(format(Error.PROVINCE_NOT_OWNED, p.name));
                    p.troops += amt;
                    troopsToDeploy -= amt;
                    t.writefln("Deployed %d troops to %s", amt, p.name);
                    t.writefln("You have %d troops remaining", troopsToDeploy);
                  } catch(ConvException e) {
                    throw new CommandException(format(Error.NOT_NUMBER, args[0]));
                  }
                  break;
                }
                case "end":
                  if(troopsToDeploy > 0) 
                    throw new CommandException(format(Error.NOT_DEPLOYED_ALL_TROOPS));
                  expectArgs(cmd, args, 0);
                  nextTurn(&t);
                  break;
                case "toggle": {
                  expectArgs(cmd, args, 1);
                  string toggle = args[0];
                  if(toggle in toggles) {
                    toggles[toggle] = !toggles[toggle];
                    t.writefln("%s is now %s", toggle, toggles[toggle]);
                  } else {
                    throw new CommandException(format(Error.NO_TOGGLE, toggle));
                  }
                  break;
                }
                case "toggles": {
                  expectArgs(cmd, args, 0);
                  t.writeln("Toggles: ");
                  t.writefln("waitBattleRounds: %s\n\tWaits 500ms for every round of a visible battle", toggles["waitBattleRounds"]);
                  t.writefln("seeAllBattles: %s\n\tDisplays all battles", toggles["seeAllBattles"]);
                  break;
                }
                case "save":
                  expectArgs(cmd, args, 1);
                  std.file.write(format("./data/saves/%s.qsf", args[0]), save(0));
                  t.writeln("Save complete");
                  break;
                case "script":
                  expectArgs(cmd, args, 1);
                  if(!toggles["cheats"]) throw new CommandException(format(Error.NO_COMMAND, cmd));
                  runScript(args[0], new GlobalScope(), &t);
                  break;
                case "run": {
                  expectAtLeastArgs(cmd, args, 1);
                  if(!toggles["cheats"]) throw new CommandException(format(Error.NO_COMMAND, cmd));
                  string script = args.join(" ");
                  try {
                    interpret(parseScript(script), new GlobalScope(), &t);
                  } catch(ScriptException e) {
                    throw new CommandException(to!string(e.message));
                  }
                  break;
                }
                default:
                  throw new CommandException(format(Error.NO_COMMAND, cmd));
              }
            }
            break;
          }
        }
      } catch(UserInterruptionException e) {
        return;
      } catch(CommandException e) {
        t.color(Color.red, Color.black);
        t.writeln(e.message);
        t.color(Color.white, Color.black);
      }
    }
  } catch(Throwable e) {
    // this exists so that program doesn't immediately close when an error occurs
    t.color(Color.white, Color.black);
    t.writeln("Error handler triggered.\n");
    t.color(Color.red, Color.black);
    t.writeln(e.toString());
    t.color(Color.white, Color.black);
    t.writeln("\nPress any key to exit.");
    auto input = RealTimeConsoleInput(&t, ConsoleInputFlags.raw);
    input.getch();
    return;
  }
}

/// Reads a command from the user and returns it as an array of strings.
string[] nextCommand(Terminal* t) {
  return t.getline().split(" ");
}

/// Advances the game by one turn.
void nextTurn(Terminal* t) {
  if(!firstTurn) {
    commitAll(actions, t);
    actions = [];
    // run AI
    foreach(ref Country c; countries) {
      if(&c != player && c.getAllProvinces().length > 0) c.ai(t);
    }
    // update
    foreach(ref Country c; countries) {
      c.update(t);
    }
    auto input = RealTimeConsoleInput(t, ConsoleInputFlags.raw);
    bool won = true;
    bool lost = true;
    foreach(Province p; provinces) {
      if(p.owner == player) lost = false;
      if(!p.ocean && p.owner != player) won = false;
    }
    if(lost) {
      t.writeln("You have been defeated!");
      renderMap(MapRenderType.COUNTRY, t);
      t.flush();
      input.getch();
      exit(0);
    } else if(won) {
      t.writeln("You have achieved world domination!");
      renderMap(MapRenderType.COUNTRY, t);
      t.flush();
      input.getch();
      exit(0);
    }
  }
  firstTurn = false;
  turn++;
  troopsToDeploy = player.troopsPerTurn();
  t.writefln("Turn %d", turn);
  t.writefln("You can deploy %d troops", troopsToDeploy);
}

/// Changes the game mode to a desired mode.
void changeMode(GameMode newMode, Terminal* t) {
  gameMode = newMode;
  final switch(gameMode) {
    case GameMode.MAIN_MENU:
      t.writeln(readText("./data/logo.txt"));
      t.writeln(readText("./data/version.txt"));
      t.writeln("Type 'help' in any mode for help");
      break;
    case GameMode.GAME:
      t.writefln("Playing as %s", player.name);
      turn = 0;
      nextTurn(t);
      break;
  } 
}